Source/WebCore/css/CSSBasicShapes.cpp

```diff
@@ -229,92 +234,115 @@ bool CSSXywhValue::equals(const CSSXywhValue& other) const
-        areDefaultCornerRadii &= buildRadii(verticalRadii, topLeftRadiusHeight, topRightRadiusHeight, bottomRightRadiusHeight, bottomLeftRadiusHeight);
+    unsigned numberOfCorners = 4;
+    for (unsigned i = 0; i < 4; ++i) {
+        auto value = extractCornerRadiusWidthAndHeight(corners[i]);
+        if (!value) {
+            numberOfCorners = i;
+            break;
+        }
 
-        if (!areDefaultCornerRadii) {
-            result.append(" round"_s);
+        horizontalRadii[i] = value->first.ptr();
+        verticalRadii[i] = value->second.ptr();
+    }
 
-            for (auto& radius : horizontalRadii)
-                result.append(' ', radius);
+    if (!numberOfCorners)
+        return;
 
-            if (verticalRadii != horizontalRadii) {
-                result.append(" /"_s);
-                for (auto& radius : verticalRadii)
-                    result.append(' ', radius);
-            }
-        }
-    }
-}
+    ASSERT(numberOfCorners == 4);
 
-static String buildXywhString(const String& insetX, const String& insetY, const String& width, const String& height,
-    const String& topLeftRadiusWidth, const String& topLeftRadiusHeight,
-    const String& topRightRadiusWidth, const String& topRightRadiusHeight,
-    const String& bottomRightRadiusWidth, const String& bottomRightRadiusHeight,
-    const String& bottomLeftRadiusWidth, const String& bottomLeftRadiusHeight)
-{
-    StringBuilder result;
-    result.append("xywh("_s, insetX, ' ', insetY, ' ', width, ' ', height);
+    auto areDefaultCornerRadii = [&](const auto (&r)[4]) {
+        if (!r[3]->equals(*r[1]))
+            return false;
+        else if (!r[2]->equals(*r[0]))
+            return false;
+        else if (!r[1]->equals(*r[0]))
+            return false;
+        else
+            return isZeroPX(*r[0]);
+    };
+
+    if (areDefaultCornerRadii(horizontalRadii) && areDefaultCornerRadii(verticalRadii))
+        return;
 
-    buildRadiiString(result, topLeftRadiusWidth, topLeftRadiusHeight, topRightRadiusWidth, topRightRadiusHeight, bottomRightRadiusWidth, bottomRightRadiusHeight, bottomLeftRadiusWidth, bottomLeftRadiusHeight);
+    builder.append(" round "_s);
 
-    result.append(')');
-    return result.toString();
+    // Beyond this point, the logic in this function is identical to that
+    // of ShorthandSerializer::serializeBorderRadius(). It would be good to
+    // find a way to share.
+
+    bool serializeBoth = false;
+    for (unsigned i = 0; i < 4; ++i) {
+        if (!horizontalRadii[i]->equals(*verticalRadii[i])) {
+            serializeBoth = true;
+            break;
+        }
+    }
+
+    auto serializeRadiiAxis = [&](const auto (&r)[4]) {
+        if (!r[3]->equals(*r[1])) {
+            r[0]->cssText(builder);
+            builder.append(' ');
+            r[1]->cssText(builder);
+            builder.append(' ');
+            r[2]->cssText(builder);
+            builder.append(' ');
+            r[3]->cssText(builder);
```

Comment:
Code like this makes it clear that an interface built more on concatenation might lead to a cleaner idiom, and even possible more efficiency since we might resize a smaller number of times. But I don’t know how to do that without code bloat.